src/multiframe.ts

```diff
@@ -0,0 +1,233 @@
+            }
+        }
+
+        // closes sample first
+        this.samples.sort((a, b) => {
+            const aL = a.length();
+            const bL = b.length();
+            return aL < bL ? -1 : (bL < aL ? 1 : 0);
+        });
+
+        const pmat = this.camera.projectionMatrix;
+        const store = new pc.Vec2();
+
+        // just before rendering the scene we apply a subpixel jitter
+        // to the camera's projection matrix.
+        this.camera.onPreRender = () => {
+            const sample = this.samples[this.sampleId];
+
+            store.set(pmat.data[12], pmat.data[13]);
+            pmat.data[8] += sample.x / device.width;
+            pmat.data[9] += sample.y / device.height;
+
+            // look away
+            this.camera._camera._viewMatDirty = true;
+            this.camera._camera._viewProjMatDirty = true;
+
+            this.textureBiasUniform.setValue(this.sampleId === 0 ? 0.0 : -5.0);
+            // this.textureBiasUniform.setValue(-5.0);
+        };
+
+        // restore the camera's projection matrix jitter once rendering is
+        // done
+        this.camera.onPostRender = () => {
+            pmat.data[8] = store.x;
+            pmat.data[9] = store.y;
+        };
+
+        this.shader = new pc.Shader(device, {
+            attributes: {
+                vertex_position: pc.SEMANTIC_POSITION
+            },
+            vshader: vertexShaderHeader(device) + vshader,
+            fshader: fragmentShaderHeader(device) + fshader
+        });
+
+        this.pixelFormat = choosePixelFormat(device);
+        this.multiframeTexUniform = device.scope.resolve('multiframeTex');
+        this.powerUniform = device.scope.resolve('power');
+        this.textureBiasUniform = device.scope.resolve('textureBias');
+
+        const handler = () => {
+            this.destroy();
+        };
+
+        device.once('destroy', handler);
+        device.on('devicelost', handler);
+    }
+
+    destroy() {
+        if (this.accumRenderTarget) {
+            this.accumRenderTarget.destroy();
+            this.accumRenderTarget = null;
+        }
+
+        if (this.accumTexture) {
+            this.accumTexture.destroy();
+            this.accumTexture = null;
+        }
+    }
+
+    create() {
+        this.accumTexture = new pc.Texture(this.device, {
+            width: this.device.width,
+            height: this.device.height,
+            format: this.pixelFormat,
+            mipmaps: false
+        });
+
+        this.accumRenderTarget = new pc.RenderTarget({
+            colorBuffer: this.accumTexture,
+            depth: false
+        });
+    }
+
+    // flag the camera as moved
+    moved() {
+        this.sampleId = 0;
+    }
+
+    // update the multiframe accumulation buffer.
+    // blend the camera's render target colour buffer with the multiframe accumulation buffer.
+    // writes results to the backbuffer.
+    update() {
+        const device = this.device as pc.WebglGraphicsDevice;
```

Comment:
I think you can get away with dropping the `as pc.WebglGraphicsDevice` now because you've updated the type of `this.device`.

Reply:
yep, done thanks!